implementations/rust/ockam/ockam_macro/src/lib.rs

```diff
@@ -0,0 +1,179 @@
+        Struct(s) => s,
+        _ => {
+            panic!("currently only works for structs");
+        }
+    };
+    let struct_ident = ast.ident;
+    let named_fields = match struct_data.fields {
+        syn::Fields::Named(f) => f.named,
+        _ => {
+            panic!("currently only works for named fields");
+        }
+    };
+
+    // Get generic type params from struct definition
+    let generic_tys = ast
+        .generics
+        .type_params()
+        .map(|t| &t.ident)
+        .collect::<Vec<_>>();
+
+    // Types for form name: T where T is a generic type
+    let simple_generic_fields = named_fields
+        .iter()
+        .filter_map(|f| {
+            let outer = get_outer(&f.ty).unwrap();
+            if generic_tys.iter().any(|id| id.to_string() == outer) {
+                return Some(outer);
+            }
+            None
+        })
+        .collect::<Vec<_>>();
+
+    // Types which have a generic and are not simple
+    let complex_generic_fields = named_fields
+        .iter()
+        .filter_map(|f| {
+            if has_generic(&f.ty, &generic_tys) && get_inner(&f.ty).is_some() {
+                return Some(&f.ty);
+            }
+            None
+        })
+        .collect::<Vec<_>>();
+
+    let mut generics = ast.generics;
+
+    // Add trait bounds on generic type params
+    for p in &mut generics.params {
+        if let GenericParam::Type(ref mut t) = *p {
+            // Every generic type must be Send + Sync
+            t.bounds.push(parse_quote!(::std::marker::Send));
+            t.bounds.push(parse_quote!(::std::marker::Sync));
+
+            // Generic simple type must also be AsyncTryClone
+            if simple_generic_fields
+                .iter()
+                .any(|s| s == &t.ident.to_string())
+            {
+                t.bounds
+                    .push(parse_quote!(ockam_core::traits::AsyncTryClone));
+            }
+        }
+    }
+
+    // Add where bounds
+    let where_clause = generics.make_where_clause();
+    for ty in complex_generic_fields {
+        where_clause
+            .predicates
+            .push(parse_quote!(#ty: ockam_core::traits::AsyncTryClone));
+    }
+
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+    let async_trait: Attribute = parse_quote!(#[async_trait]);
+    let fields = named_fields.iter().map(|f| {
+        let field_name = &f.ident;
+        quote! {
+            #field_name
+        }
+    });
+    let fields_clone = fields.clone();
+    let fields_async_impls = named_fields.iter().map(|f| {
+        let field_name = &f.ident;
+        quote! {
+            self.#field_name.async_try_clone()
+        }
+    });
+    let trait_fn = quote! {
+        async fn async_try_clone(&self) -> ockam_core::Result<Self>{
+            let results = ockam_core::try_join!(
+                #(#fields_async_impls),*
+            );
+            match results {
+                Ok((#(#fields),*))=> {
```

Comment:
This doesn't work for single-field struct, because in this case parentheses won't map tuple into individual field. The possible solution could be take tuple here and iterate over it. Maybe there are cleaner ways to achieve it.

Reply:
Another possible solution would be to append a comma after the expansion which would work for both single-field and multi-field structs.